tests/Tests.Shared/test-api.h

```diff
@@ -0,0 +1,599 @@
+
+#ifdef EXPECT_STRCASENE_MSG
+#undef EXPECT_STRCASENE_MSG
+#endif
+#define EXPECT_STRCASENE_MSG(expected, actual, format, ...)                \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_STRCASENE(expected, actual);
+
+#ifdef EXPECT_FLOAT_EQ_MSG
+#undef EXPECT_FLOAT_EQ_MSG
+#endif
+#define EXPECT_FLOAT_EQ_MSG(expected, actual, format, ...)                 \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_FLOAT_EQ(expected, actual);
+
+#ifdef EXPECT_DOUBLE_EQ_MSG
+#undef EXPECT_DOUBLE_EQ_MSG
+#endif
+#define EXPECT_DOUBLE_EQ_MSG(expected, actual, format, ...)                \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_DOUBLE_EQ(expected, actual);
+
+#ifdef EXPECT_NEAR_MSG
+#undef EXPECT_NEAR_MSG
+#endif
+#define EXPECT_NEAR_MSG(val1, val2, abs_error, format, ...)                \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_NEAR(val1, val2, abs_error);
+
+#ifdef EXPECT_THROW_MSG
+#undef EXPECT_THROW_MSG
+#endif
+#define EXPECT_THROW_MSG(statement, expected_exception, format, ...)       \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_EXPECT_THROW(statement, expected_exception);
+
+#ifdef EXPECT_NO_THROW_MSG
+#undef EXPECT_NO_THROW_MSG
+#endif
+#define EXPECT_NO_THROW_MSG(statement, format, ...)                        \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_NO_THROW(statement);
+
+#ifdef EXPECT_ANY_THROW_MSG
+#undef EXPECT_ANY_THROW_MSG
+#endif
+#define EXPECT_ANY_THROW_MSG(statement, format, ...)                       \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_ANY_THROW(statement);
+
+#ifdef EXPECT_OBJCEQ_MSG
+#undef EXPECT_OBJCEQ_MSG
+#endif
+#define EXPECT_OBJCEQ_MSG(val1, val2, format, ...)                         \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_OBJCEQ(val1, val2);
+
+#ifdef EXPECT_OBJCNE_MSG
+#undef EXPECT_OBJCNE_MSG
+#endif
+#define EXPECT_OBJCNE_MSG(val1, val2, format, ...)                         \
+    TestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
+    EXPECT_OBJCNE(val1, val2);
+
+#ifdef FRIEND_TEST
+#undef FRIEND_TEST
+#endif
+#define FRIEND_TEST(test_case_name, test_name) friend class test_case_name##_##test_name
+
+// clang-format off
+template <unsigned NumChar>
+constexpr const wchar_t* const _IsDisabledTest(char const (&str1)[NumChar]) {
+    return ((NumChar > 8)  &&
+            str1[0] == 'D' &&
+            str1[1] == 'I' &&
+            str1[2] == 'S' &&
+            str1[3] == 'A' &&
+            str1[4] == 'B' &&
+            str1[5] == 'L' &&
+            str1[6] == 'E' &&
+            str1[7] == 'D') ? L"TRUE" : L"FALSE";
+}
+
+__declspec(noinline) inline std::string _ExchangeAlternateTestName(
```

Comment:
extra special inlining. It means __declspec(selectany) for inline functions. 

https://blogs.msdn.microsoft.com/freik/2005/10/26/how-to-make-a-no-inline-inline-function-and-why-you-might-want-to/ has a good summary. 

The gist of it is that I'm using a global variable that needs to be shared across TUs (DrawingTest and the Test file that derives from it). This means I *can't* have more than one static function stamped out / use an inline function because the storage for the alternate name wouldn't be the same across TUs. However, I wanted to keep this in the header to minimize the additional goop needed by the tests (an alternate solution would be to place this stuff in the logger static lib). This means I need a "selectAny" type of behavior which you get from this hilarious syntax.